### ISSUE TITLE
[HF transformers] ReformerModel LSH hashing non-deterministic under torch.compile(backend="eager") — config.hash_seed not honored, breaks bitwise equivalence

### ISSUE THREAD

**Issue by penguinwu** [penguinwu/oss-model-graph-break-corpus#131]
## Summary

`ReformerModel` (and other Reformer variants) produce **different output under `torch.compile(model, backend="eager", fullgraph=False)` vs eager** (max abs diff ~3–5). Root cause: the LSH attention's random rotations are drawn from the **global RNG**, and under `torch.compile` Dynamo substitutes its own RNG — so the rotations, and therefore the LSH bucket assignment, differ from eager, producing O(1) output divergence.

Critically, HF's built-in determinism hook for this — **`config.hash_seed`** (which runs `torch.manual_seed(self.hash_seed)` before the draw) — does **not** make it deterministic under `torch.compile`, because Dynamo does not honor an in-graph global `manual_seed`. (Verified: setting `hash_seed` does not change the divergence under compile.)

This is the companion to the SDPA (#129) and LayerDrop/dropout-RNG (#130) bitwise-equivalence issues; Reformer is the remaining model class in our sweep whose `backend="eager"` output diverges from eager.

## Environment

- transformers 5.9.0 (current latest; the code path is unchanged across recent releases)
- torch 2.13 nightly (cu126); reproduces on stable torch ≥ 2.3
- CUDA GPU

## Reproduction

```python
import torch, torch._dynamo
from transformers import ReformerModel, ReformerConfig

# Minimal self-contained config: seq_len = prod(axial_pos_shape) = 32; hidden = sum(axial_pos_embds_dim)
cfg = ReformerConfig(
    attn_layers=["lsh", "lsh"], num_hashes=2, num_buckets=4, lsh_attn_chunk_length=8,
    hidden_size=64, feed_forward_size=128, num_attention_heads=2, attention_head_size=16,
    axial_pos_shape=[4, 8], axial_pos_embds_dim=[32, 32], vocab_size=320,
    hidden_dropout_prob=0.1, lsh_attention_probs_dropout_prob=0.1,
)
torch.manual_seed(0)
model = ReformerModel(cfg).cuda().train()
x = torch.randint(0, 320, (1, 32)).cuda()

def fwd(m):
    torch._dynamo.reset(); torch.manual_seed(0)
    with torch.enable_grad():
        return m(x).last_hidden_state

oe = fwd(model)
oc = fwd(torch.compile(model, backend="eager", fullgraph=False))
print("bitwise_equal:", torch.equal(oe.detach().cpu(), oc.detach().cpu()))  # False
print("max_diff:", (oe.detach() - oc.detach()).abs().max().item())          # ~3.5
```

## Root cause

`transformers/models/reformer/modeling_reformer.py::LSHSelfAttention._hash_vectors` (~L713–719):

```python
if self.hash_seed is not None:
    # for determinism
    torch.manual_seed(self.hash_seed)

rotations_shape = (self.num_attention_heads, vectors.shape[-1], num_hashes, rotation_size // 2)
random_rotations = torch.randn(rotations_shape, device=vectors.device, dtype=vectors.dtype)
```

- **Eager:** `random_rotations` comes from the global RNG; with `hash_seed` set, `manual_seed` makes it deterministic.
- **Compile:** Dynamo functionalizes/substitutes the RNG for `torch.randn`, and the in-graph `torch.manual_seed` is not honored, so the compiled rotations differ from eager → different LSH buckets → divergent output. **`config.hash_seed` does not fix this under compile.**

It is **not** a non-deterministic CUDA kernel — `torch.use_deterministic_algorithms(True)` passes clean on the forward+backward; the divergence is purely the RNG-sourced rotations.

(Secondary, related: in train mode Reformer also applies `nn.functional.dropout(x, p=self.dropout)` where `self.dropout` is a dropout prob cached at `__init__` — this is the same `torch.compile` RNG-divergence class as #130 and stacks on top of the rotation issue. Disabling train-mode dropout isolates the rotation effect.)

## Proposed fix

Draw the rotations from an explicit **`torch.Generator`** seeded by `hash_seed`, instead of relying on the global `torch.manual_seed`. An explicit generator's state is independent of Dynamo's global-RNG substitution, so eager and compiled paths draw identical rotations:

```python
gen = torch.Generator(device=vectors.device)
gen.manual_seed(self.hash_seed if self.hash_seed is not None else SOME_DEFAULT)
random_rotations = torch.randn(
    rotations_shape, device=vectors.device, dtype=vectors.dtype, generator=gen
)
```

This makes `config.hash_seed` actually deliver determinism under `torch.compile` (its documented purpose) while leaving default eager behavior unchanged. The exact generator lifecycle is a maintainer call — e.g. persist the generator on the module and let it advance across calls/rounds to preserve per-round rotation diversity, rather than re-seeding identically each call.

## Validation

With rotations drawn from an explicit generator (and train-mode dropout disabled to isolate the rotation effect), `ReformerModel` in train mode becomes **bitwise-equal** between eager and `torch.compile(backend="eager", fullgraph=False)`: `torch.equal(...) == True`, max_diff `0.0` (verified on tf 5.9.0). Without the fix, max_diff ~3.5–4.6.
